lib/simulator_sse.h

```diff
@@ -138,88 +137,68 @@ class SimulatorSSE final {
+    __m128 u[4];
+
+    switch (q0) {
+    case 0:
+      u[0] = SetPs(matrix, 6, 0, 6, 0);
+      u[1] = SetPs(matrix, 7, 1, 7, 1);
+      u[2] = SetPs(matrix, 4, 2, 4, 2);
+      u[3] = SetPs(matrix, 5, 3, 5, 3);
+      break;
+    case 1:
+      u[0] = SetPs(matrix, 6, 6, 0, 0);
+      u[1] = SetPs(matrix, 7, 7, 1, 1);
+      u[2] = SetPs(matrix, 4, 4, 2, 2);
+      u[3] = SetPs(matrix, 5, 5, 3, 3);
+      break;
+    default:
+      // Cannot reach here.
+      for (std::size_t i = 0; i < 4; ++i) {
+        u[i] = _mm_set1_ps(0);
+      }
+      break;
+    }
+
     uint64_t sizei = uint64_t{1} << (num_qubits_ + 1);
 
     fp_type* rstate = StateSpace::RawData(state);
 
     auto f = [](unsigned n, unsigned m, uint64_t i, unsigned q0,
-                const fp_type* matrix, fp_type* rstate) {
-      __m128 r0, i0, r1, i1, ru, iu, rn, in, rm, im;
-
-      uint64_t p = 8 * i;
-
-      r0 = _mm_load_ps(rstate + p);
-      i0 = _mm_load_ps(rstate + p + 4);
-
-      switch (q0) {
-      case 0:
-        r1 = _mm_shuffle_ps(r0, r0, 49);  // 00110001
-        i1 = _mm_shuffle_ps(i0, i0, 49);
-        break;
-      case 1:
-        r1 = _mm_shuffle_ps(r0, r0, 14);  // 00001110
-        i1 = _mm_shuffle_ps(i0, i0, 14);
-        break;
-      default:
-        // Cannot reach here.
-        break;
-      }
+                const __m128* u, fp_type* rstate) {
+      __m128 r0, i0, r1, i1, rn, in;
 
-      ru = _mm_set1_ps(matrix[0]);
-      iu = _mm_set1_ps(matrix[1]);
-      rn = _mm_mul_ps(r0, ru);
-      in = _mm_mul_ps(r0, iu);
-      rn = _mm_sub_ps(rn, _mm_mul_ps(i0, iu));
-      in = _mm_add_ps(in, _mm_mul_ps(i0, ru));
-      ru = _mm_set1_ps(matrix[2]);
-      iu = _mm_set1_ps(matrix[3]);
-      rn = _mm_add_ps(rn, _mm_mul_ps(r1, ru));
-      in = _mm_add_ps(in, _mm_mul_ps(r1, iu));
-      rn = _mm_sub_ps(rn, _mm_mul_ps(i1, iu));
-      in = _mm_add_ps(in, _mm_mul_ps(i1, ru));
+      auto p0 = rstate + 8 * i;
 
-      ru = _mm_set1_ps(matrix[4]);
-      iu = _mm_set1_ps(matrix[5]);
-      rm = _mm_mul_ps(r0, ru);
-      im = _mm_mul_ps(r0, iu);
-      rm = _mm_sub_ps(rm, _mm_mul_ps(i0, iu));
-      im = _mm_add_ps(im, _mm_mul_ps(i0, ru));
-      ru = _mm_set1_ps(matrix[6]);
-      iu = _mm_set1_ps(matrix[7]);
-      rm = _mm_add_ps(rm, _mm_mul_ps(r1, ru));
-      im = _mm_add_ps(im, _mm_mul_ps(r1, iu));
-      rm = _mm_sub_ps(rm, _mm_mul_ps(i1, iu));
-      im = _mm_add_ps(im, _mm_mul_ps(i1, ru));
-
-      switch (q0) {
-      case 0:
-        rm = _mm_shuffle_ps(rm, rm, 128);  // 10000000
-        im = _mm_shuffle_ps(im, im, 128);
-        rn = _mm_blend_ps(rn, rm, 10);  // 1010
-        in = _mm_blend_ps(in, im, 10);
-        break;
-      case 1:
-        rn = _mm_shuffle_ps(rn, rm, 68);  // 01000100
-        in = _mm_shuffle_ps(in, im, 68);
-        break;
-      default:
-        // Cannot reach here.
-        break;
-      }
+      // 177 = 0b10110001, 78 = 0b01001110
```

Comment:
Could you add context to this comment or remove it?

Reply:
These numbers specify how to shuffle four elements in `__m128` registers. I added a short explanation.